### PR TITLE
syncer: replace `{schema, table}` with `filter.Table` Part 3

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve? 

Part of https://github.com/pingcap/dm/issues/2078 and https://github.com/pingcap/dm/issues/2064

### What is changed and how it works?

- Replace `{schema, table}` with `Table`.

### Check List 

Tests 

 - unit test

Code changes

 - Has interface methods change

### CODE REVIEW

File: syncer/ddl.go [hunk 1]
```diff
@@ -86,12 +86,12 @@ func (s *Syncer) parseDDLSQL(sql string, p *parser.Parser, schema string) (resul
 		}, nil
 	case ast.DMLNode:
 		// if DML can be ignored, we do not report an error
-		schema2, table, err2 := tableNameForDML(node)
+		table, err2 := tableNameForDML(node)
 		if err2 == nil {
-			if len(schema2) > 0 {
-				schema = schema2
+			if len(table.Schema) == 0 {
+				table.Schema = schema
```
Comment (Ehco1996) [inline syncer/ddl.go:92]: this change seems inconsistent with old logic?

File: syncer/ddl.go [hunk 2]
```diff
@@ -312,5 +312,7 @@ func (s *Syncer) clearOnlineDDL(tctx *tcontext.Context, targetTable *filter.Tabl
 type shardingDDLInfo struct {
 	name   string
 	tables [][]*filter.Table
-	stmt   ast.StmtNode
+	// sourceTables []*filter.Table
```
Comment (Ehco1996) [inline syncer/ddl.go:315]: why comment this?

File: syncer/ddl.go (hunk 1 above)
Comment (okJiang) [reply syncer/ddl.go:92]: I don't think so. Please check it again.

File: syncer/ddl.go (hunk 2 above)
Comment (okJiang) [reply syncer/ddl.go:315]: This will not commit this pr. I will remove it next commit.

BTW, I think it has two improvements:
1. `tables [][]*filter.Table` can split to `sourceTables []*filter.Table` and `targetTables []*filter.Table`;
2. In some cases, `tables [][]*filter.Table` can change to `tables []*filter.Table`, and you can see why in https://github.com/pingcap/dm/issues/2129

We can discuss it in https://github.com/pingcap/dm/issues/2129

File: syncer/ddl.go (hunk 1 above)
Comment (Ehco1996) [reply syncer/ddl.go:92]: old: use schema from ast node 
new: always use the schema passed in 🤔

File: syncer/ddl.go (hunk 2 above)
Comment (Ehco1996) [reply syncer/ddl.go:315]: SGTM

File: syncer/ddl.go (hunk 1 above)
Comment (okJiang) [reply syncer/ddl.go:92]: only use `schema` in 

```go
if len(table.Schema) == 0 {
```

File: syncer/ddl_test.go [hunk 3]
```diff
@@ -496,12 +496,12 @@ func (m mockOnlinePlugin) Apply(tctx *tcontext.Context, tables []*filter.Table,
 	return nil, nil
 }
 
-func (m mockOnlinePlugin) Finish(tctx *tcontext.Context, schema, table string) error {
-	k := schema + table
-	if _, ok := m.toFinish[k]; !ok {
+func (m mockOnlinePlugin) Finish(tctx *tcontext.Context, table *filter.Table) error {
+	tableID := utils.GenTableID(table)
```
Comment (glorv) [inline syncer/ddl_test.go:500]: Does the caller also need to change size the key pattern has changed?

File: syncer/dml.go [hunk 4]
```diff
@@ -32,8 +32,7 @@ import (
 
 // genDMLParam stores pruned columns, data as well as the original columns, data, index.
 type genDMLParam struct {
-	schema            string
-	table             string
+	tableID           string
```
Comment (glorv) [inline syncer/dml.go:35]: Please add a comment to explain the pattern of tableID

File: syncer/dml.go [hunk 5]
```diff
@@ -569,9 +568,9 @@ func (s *Syncer) mappingDML(schema, table string, ti *model.TableInfo, data [][]
 		rows = make([][]interface{}, len(data))
 	)
 	for i := range data {
-		rows[i], _, err = s.columnMapping.HandleRowValue(schema, table, columns, data[i])
+		rows[i], _, err = s.columnMapping.HandleRowValue(table.Schema, table.Name, columns, data[i])
```
Comment (glorv) [inline syncer/dml.go:571]: Why not also change `HandleRowValue` to accept a `filter.Table`

File: syncer/job.go [hunk 6]
```diff
@@ -64,9 +65,8 @@ type job struct {
 	// ddl in ShardOptimistic and ShardPessimistic will only affect one table at one time but for normal node
 	// we don't have this limit. So we should update multi tables in normal mode.
 	// sql example: drop table `s1`.`t1`, `s2`.`t2`.
-	sourceTbl       map[string][]string
-	targetSchema    string
-	targetTable     string
+	sourceTbls      map[string][]string
```
Comment (glorv) [inline syncer/job.go:68]: Seems change this type to `map[string].filter.Table` can make the code a little simpler

**Review by glorv — COMMENTED** [review]
rest LGTM

File: syncer/ddl_test.go (hunk 3 above)
Comment (okJiang) [reply syncer/ddl_test.go]: yes, I changed it back.

File: syncer/dml.go (hunk 5 above)
Comment (okJiang) [reply syncer/dml.go:571]: It belongs to `tidb-tools`.

File: syncer/job.go (hunk 6 above)
Comment (okJiang) [reply syncer/job.go:68]: Should be `map[string]*filter.Table` --> `map[string][]*filter.Table`.

Comment (okJiang) [reply syncer/job.go:68]: And I have thought change it to `[]*filter.Table`, but dismiss it because it will affect performance in https://github.com/pingcap/dm/blob/master/syncer/syncer.go#L1032-L1051

Comment (glorv) [reply syncer/job.go]: I don't understand why it will hurt performance since L1032-1051 also need filter.Table instead of separate shcema/table name, with this change we can avoid newing the structs

File: syncer/dml.go (hunk 5 above)
Comment (glorv) [reply syncer/dml.go:571]: 😂ok.

File: syncer/job.go (hunk 6 above)
Comment (okJiang) [reply syncer/job.go]: I mean `change it to []*filter.Table` will hurt performance, because it can't directly `continue` all `len(sourceSchema) == 0`.

Comment (Ehco1996) [reply syncer/job.go]: i also think `map[string]*filter.Table` is enough here, not sure why we need to use `map[string][]*filter.Table`

Comment (okJiang) [reply syncer/job.go]: Because one `Schema` may map lots of `Table`s.

Comment (Ehco1996) [reply syncer/job.go]: i see, we need use `map[string][]*filter.Table` to handle `multi schema change` ddl

**Review by Ehco1996 — APPROVED** [review]
LGTM

File: syncer/util.go [hunk 7]
```diff
@@ -47,41 +47,41 @@ func binlogTypeToString(binlogType BinlogType) string {
...
 
-// tableNameForDML gets table name from INSERT/UPDATE/DELETE statement.
-func tableNameForDML(dml ast.DMLNode) (schema, table string, err error) {
+// getTableByDML gets table from INSERT/UPDATE/DELETE statement.
+func getTableByDML(dml ast.DMLNode) (*filter.Table, error) {
 	switch stmt := dml.(type) {
 	case *ast.InsertStmt:
 		if stmt.Table == nil || stmt.Table.TableRefs == nil || stmt.Table.TableRefs.Left == nil {
-			return "", "", terror.ErrSyncUnitInvalidTableName.Generate(fmt.Sprintf("INSERT statement %s not valid", stmt.Text()))
+			return &filter.Table{}, terror.ErrSyncUnitInvalidTableName.Generate(fmt.Sprintf("INSERT statement %s not valid", stmt.Text()))
```
Comment (glorv) [inline syncer/util.go:55]:
```suggestion
			return nil, terror.ErrSyncUnitInvalidTableName.Generate(fmt.Sprintf("INSERT statement %s not valid", stmt.Text()))
```

Comment (glorv) [reply syncer/util.go:55]: If err is not nil, return nil is enough. Please also replace all the following ones.

**Review by glorv — APPROVED** [review]

**Comment by Ehco1996** [comment]
```

[2021-09-17T06:39:50.560Z] PANIC: util_test.go:29: testUtilSuite.TestGetTableByDML

[2021-09-17T06:39:50.560Z] 

[2021-09-17T06:39:50.560Z] ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x138FB86)

[2021-09-17T06:39:50.560Z] 

[2021-09-17T06:39:50.560Z] /usr/local/go/src/runtime/panic.go:971

[2021-09-17T06:39:50.560Z]   in gopanic

[2021-09-17T06:39:50.560Z] /usr/local/go/src/runtime/panic.go:212

[2021-09-17T06:39:50.560Z]   in panicmem

[2021-09-17T06:39:50.560Z] /usr/local/go/src/runtime/signal_unix.go:734

[2021-09-17T06:39:50.560Z]   in sigpanic

[2021-09-17T06:39:50.560Z] util_test.go:86

[2021-09-17T06:39:50.560Z]   in testUtilSuite.TestGetTableByDML

[2021-09-17T06:39:50.560Z] /usr/local/go/src/reflect/value.go:337

[2021-09-17T06:39:50.560Z]   in Value.Call

[2021-09-17T06:39:50.560Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850

[2021-09-17T06:39:50.560Z]   in suiteRunner.forkTest.func1

[2021-09-17T06:39:50.560Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739

[2021-09-17T06:39:50.560Z]   in suiteRunner.forkCall.func1

[2021-09-17T06:39:50.560Z] /usr/local/go/src/runtime/asm_amd64.s:1371

[2021-09-17T06:39:50.560Z]   in goexit

```
panic in test 😂 cc @okJiang

File: tests/online_ddl/run.sh [hunk 8]
```diff
@@ -140,4 +140,4 @@ cleanup_process $*
 run
 cleanup_process $*
 
-echo "[$(date)] <<<<<< test case $TEST_NAME success! >>>>>>"
+echo "[$(date)] <<<<<< test case $TEST_NAME success! >>>>>>"
```
Comment (Ehco1996) [inline tests/online_ddl/run.sh:143]: please revert this change
